api/v3/Generic/Getlist.php

```diff
@@ -37,6 +38,23 @@ function civicrm_api3_generic_getList($apiRequest) {
 
   $request['params']['check_permissions'] = !empty($apiRequest['params']['check_permissions']);
   $result = civicrm_api3($entity, 'get', $request['params']);
+  if (!empty($request['input']) && !empty($defaults['search_field_fallback']) && $result['count'] < $request['params']['options']['limit']) {
+    // We support a field fallback. Note we don't do this as an OR query because that could easily
+    // bypass an index & kill the server. We just 'pad' the results if needed with the second
+    // query - this is effectively the same as what the old Ajax::getContactEmail function did.
+    // Since these queries should be quick & often only one should be needed this is a simpler alternative
+    // to constructing a UNION via the api.
+    $request['params'][$defaults['search_field_fallback']] = $request['params'][$defaults['search_field']];
+    unset($request['params'][$defaults['search_field']]);
+    $request['params']['options']['limit'] -= $result['count'];
+    $result2 = civicrm_api3($entity, 'get', $request['params']);
```

Comment:
Maybe we should add something to the params here that excludes the original results.
`NOT IN(array_keys($results))` would work partially, but it would only exclude the current page.
What about a "NOT LIKE" clause to exclude any results that would have been returned by searching on the original field.
The reason I bring this up is because we are dealing with a paging situation so duplicate results can mess up the pager.

Reply:
@colemanw - that makes sense. In my testing it had no negative performance impact adding that - however my testing did demonstrate that the  way  it was doing the ordering WAS having a performance impact and that for  this to perform well the sort_field needs to be the same as the filter field. I'm on the fence about  doing a post-sort in php - as of now its


1) sort_name matches, ordered by sort_name, labeled by sort_name padded with
2) email matches, ordered by email, labeled by sort_name

Reply:
@eileenmcnaughton this is looking really good. My one hesitation is that the `'NOT IN'` approach will possibly return redundant results because of the pager. Ex:
- Type a search query
- It returns 10 results.
- Scroll to the end and another 10 results appears (page 2)
- Scroll down again and 5 results are fetched by the main `$request`. It then fires the fallback request, using `'NOT IN'` to exclude the 5 current results.
- Since it did not exclude results from pages 1 & 2 (only the five contacts from page 3 were excluded) you'll get duplicates on page 3+.

If it's not a big performance hit, the `'NOT LIKE'` method could fix this.

Reply:
Well it was OK in the performance tests below so trying it